trieste/acquisition/function.py

```diff
@@ -745,6 +745,56 @@ def ehvi_based_on_partitioned_cell(
     return acquisition
 
 
+class ExpectedConstrainedHypervolumeImprovement(ExpectedConstrainedImprovement):
+    """
+    Builder for the constrained expected hypervolume improvement acquisition function.
+    This function essentially combines ExpectedConstrainedImprovement and ExpectedHypervolumeImprovement.
+    """
+
+    def __repr__(self) -> str:
+        """"""
+        return (
+            f"ExpectedConstrainedHypervolumeImprovement({self._objective_tag!r}, {self._constraint_builder!r},"
+            f" {self._min_feasibility_probability!r})"
+        )
+
+    def prepare_acquisition_function(
+        self, datasets: Mapping[str, Dataset], models: Mapping[str, ProbabilisticModel]
+    ) -> AcquisitionFunction:
+        """
+        :param datasets: The data from the observer. Must be populated.
+        :param models: The models over each dataset in ``datasets``.
+        :return: The expected constrained hypervolume improvement acquisition function. This function will raise
+            :exc:`ValueError` or :exc:`~tf.errors.InvalidArgumentError` if used with a batch size
+            greater than one.
+        :raise KeyError: If `objective_tag` is not found in ``datasets`` and ``models``.
+        :raise ValueError: If the objective data is empty.
+        """
+
+        objective_model = models[self._objective_tag]
+        objective_dataset = datasets[self._objective_tag]
+
+        if len(objective_dataset) == 0:
+            raise ValueError(
+                "Expected hypervolume improvement is defined with respect to existing points in the objective"
+                " data, but the objective data is empty."
+            )
+
+        constraint_fn = self._constraint_builder.prepare_acquisition_function(datasets, models)
+        pof = constraint_fn(objective_dataset.query_points[:, None, ...])
+        is_feasible = pof >= self._min_feasibility_probability
+
+        if not tf.reduce_any(is_feasible):
+            return constraint_fn
+
+        objective_mean, _ = objective_model.predict(objective_dataset.query_points)
+
+        _pf = Pareto(tf.boolean_mask(objective_mean, is_feasible[:, 0]))
```

Comment:
see my PR #284 . With that, you can a) make this faster, and b) merge this with `ExpectedConstrainedImprovement` if you so wish

Reply:
I've merged @joelberkeley's PR #284 . So maybe you can do as he suggests @vpicheny ?

Reply:
sure, will do!

Reply:
@joelberkeley what do you have in mind with merging? 

Something like 
`constraint_fn = self._constraint_builder.prepare_acquisition_function(datasets, models)` 
but for the objective?

Reply:
sorry, i missed this. I've looked into this more and I don't know how it's possible. You may just need two versions, though you could pull the first half of the method into a module-private function which both methods call if it's possible to write that in a way that's worth it